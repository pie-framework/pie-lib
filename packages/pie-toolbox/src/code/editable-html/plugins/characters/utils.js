export const spanishConfig = {
  characters: [
    ['á', 'é', 'í', 'ó', 'ú'],
    ['Á', 'É', 'Í', 'Ó', 'Ú'],
    ['—', '«', '»', 'ñ', 'ü'],
    ['-', '¿', '¡', 'Ñ', 'Ü'],
  ],
};

export const specialConfig = {
  hasPreview: true,
  characters: [
    [
      {
        unicode: 'U+00A2',
        description: 'CENT SIGN',
        write: '¢',
        label: '¢',
      },
      {
        unicode: 'U+00BF',
        description: 'INVERTED QUESTION MARK',
        write: '¿',
        label: '¿',
      },
      {
        unicode: 'U+00B4',
        description: 'ACUTE ACCENT',
        write: '´',
        label: '´',
        extraProps: { style: { gridRow: 'span 2' } },
      },
      {
        unicode: 'U+00E1',
        description: 'LATIN SMALL LETTER A WITH ACUTE',
        write: 'á',
        label: 'á',
      },
      {
        unicode: 'U+00E9',
        description: 'LATIN SMALL LETTER E WITH ACUTE',
        write: 'é',
        label: 'é',
      },
      {
        unicode: 'U+00ED',
        description: 'LATIN SMALL LETTER I WITH ACUTE',
        write: 'í',
        label: 'í',
      },
      {
        unicode: 'U+00F3',
        description: 'LATIN SMALL LETTER O WITH ACUTE',
        write: 'ó',
        label: 'ó',
      },
      {
        unicode: 'U+00FA',
        description: 'LATIN SMALL LETTER U WITH ACUTE',
        write: 'ú',
        label: 'ú',
      },
      {
        unicode: 'U+00F1',
        description: 'LATIN SMALL LETTER N WITH TILDE',
        write: 'ñ',
        label: 'ñ',
      },
    ],
    [
      {
        unicode: 'U+20AC',
        description: 'EURO SIGN',
        write: '€',
        label: '€',
      },
      {
        unicode: 'U+00A1',
        description: 'INVERTED EXCLAMATION MARK',
        write: '¡',
        label: '¡',
      },
      {
        unicode: 'U+00C1',
        description: 'LATIN CAPITAL LETTER A WITH ACUTE',
        write: 'Á',
        label: 'Á',
      },
      {
        unicode: 'U+00C9',
        description: 'LATIN CAPITAL LETTER E WITH ACUTE',
        write: 'É',
        label: 'É',
      },
      {
        unicode: 'U+00CD',
        description: 'LATIN CAPITAL LETTER I WITH ACUTE',
        write: 'Í',
        label: 'Í',
      },
      {
        unicode: 'U+00D3',
        description: 'LATIN CAPITAL LETTER O WITH ACUTE',
        write: 'Ó',
        label: 'Ó',
      },
      {
        unicode: 'U+00DA',
        description: 'LATIN CAPITAL LETTER U WITH ACUTE',
        write: 'Ú',
        label: 'Ú',
      },
      {
        unicode: 'U+00D1',
        description: 'LATIN CAPITAL LETTER N WITH TILDE',
        write: 'Ñ',
        label: 'Ñ',
      },
    ],
    [
      {
        unicode: 'U+00A3',
        description: 'POUND SIGN',
        write: '£',
        label: '£',
      },
      {
        unicode: 'U+00AB',
        description: 'LEFT-POINTING DOUBLE ANGLE QUOTATION MARK',
        write: '«',
        label: '«',
      },
      {
        unicode: 'U+005E',
        description: 'CIRCUMFLEX ACCENT',
        write: '^',
        label: '^',
        extraProps: { style: { gridRow: 'span 2' } },
      },
      {
        unicode: 'U+00E2',
        description: 'LATIN SMALL LETTER A WITH CIRCUMFLEX',
        write: 'â',
        label: 'â',
      },
      {
        unicode: 'U+00EA',
        description: 'LATIN SMALL LETTER E WITH CIRCUMFLEX',
        write: 'ê',
        label: 'ê',
      },
      {
        unicode: 'U+00EE',
        description: 'LATIN SMALL LETTER I WITH CIRCUMFLEX',
        write: 'î',
        label: 'î',
      },
      {
        unicode: 'U+00F4',
        description: 'LATIN SMALL LETTER O WITH CIRCUMFLEX',
        write: 'ô',
        label: 'ô',
      },
      {
        unicode: 'U+00FB',
        description: 'LATIN SMALL LETTER U WITH CIRCUMFLEX',
        write: 'û',
        label: 'û',
      },
      {
        unicode: 'U+00E7',
        description: 'LATIN SMALL LETTER C WITH CEDILLA',
        write: 'ç',
        label: 'ç',
      },
    ],
    [
      {
        unicode: 'U+00A5',
        description: 'YEN SIGN',
        write: '¥',
        label: '¥',
      },
      {
        unicode: 'U+00BB',
        description: 'RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK',
        write: '»',
        label: '»',
      },
      {
        unicode: 'U+00C2',
        description: 'LATIN CAPITAL LETTER A WITH CIRCUMFLEX',
        write: 'Â',
        label: 'Â',
      },
      {
        unicode: 'U+00CA',
        description: 'LATIN CAPITAL LETTER E WITH CIRCUMFLEX',
        write: 'Ê',
        label: 'Ê',
      },
      {
        unicode: 'U+00CE',
        description: 'LATIN CAPITAL LETTER I WITH CIRCUMFLEX',
        write: 'Î',
        label: 'Î',
      },
      {
        unicode: 'U+00D4',
        description: 'LATIN CAPITAL LETTER O WITH CIRCUMFLEX',
        write: 'Ô',
        label: 'Ô',
      },
      {
        unicode: 'U+00DB',
        description: 'LATIN CAPITAL LETTER U WITH CIRCUMFLEX',
        write: 'Û',
        label: 'Û',
      },
      {
        unicode: 'U+00C7',
        description: 'LATIN CAPITAL LETTER C WITH CEDILLA',
        write: 'Ç',
        label: 'Ç',
      },
    ],
    [
      {
        unicode: 'U+00A0',
        description: 'NO-BREAK SPACE',
        write: String.fromCodePoint('0x00A0'),
        label: '&nbsp;',
      },
      {
        unicode: 'U+00A7',
        description: 'SECTION SIGN',
        write: '§',
        label: '§',
      },
      {
        unicode: 'U+00A8',
        description: 'DIAERESIS',
        write: '¨',
        label: '¨',
        extraProps: { style: { gridRow: 'span 2' } },
      },
      {
        unicode: 'U+00E4',
        description: 'LATIN SMALL LETTER A WITH DIAERESIS',
        write: 'ä',
        label: 'ä',
      },
      {
        unicode: 'U+00EB',
        description: 'LATIN SMALL LETTER E WITH DIAERESIS',
        write: 'ë',
        label: 'ë',
      },
      {
        unicode: 'U+00EF',
        description: 'LATIN SMALL LETTER I WITH DIAERESIS',
        write: 'ï',
        label: 'ï',
      },
      {
        unicode: 'U+00F6',
        description: 'LATIN SMALL LETTER O WITH DIAERESIS',
        write: 'ö',
        label: 'ö',
      },
      {
        unicode: 'U+00FC',
        description: 'LATIN SMALL LETTER U WITH DIAERESIS',
        write: 'ü',
        label: 'ü',
      },
      {
        unicode: 'U+00DF',
        description: 'LATIN SMALL LETTER SHARP S',
        write: 'ß',
        label: 'ß',
      },
    ],
    [
      {
        unicode: 'U+2009',
        description: 'THIN SPACE',
        write: String.fromCodePoint('0x2009'),
        label: '&thinsp;',
      },
      {
        unicode: 'U+2026',
        description: 'HORIZONTAL ELLIPSIS',
        write: '…',
        label: '…',
      },
      {
        unicode: 'U+00C4',
        description: 'LATIN CAPITAL LETTER A WITH DIAERESIS',
        write: 'Ä',
        label: 'Ä',
      },
      {
        unicode: 'U+00CB',
        description: 'LATIN CAPITAL LETTER E WITH DIAERESIS',
        write: 'Ë',
        label: 'Ë',
      },
      {
        unicode: 'U+00CF',
        description: 'LATIN CAPITAL LETTER I WITH DIAERESIS',
        write: 'Ï',
        label: 'Ï',
      },
      {
        unicode: 'U+00D6',
        description: 'LATIN CAPITAL LETTER O WITH DIAERESIS',
        write: 'Ö',
        label: 'Ö',
      },
      {
        unicode: 'U+00DC',
        description: 'LATIN CAPITAL LETTER U WITH DIAERESIS',
        write: 'Ü',
        label: 'Ü',
      },
      {
        unicode: 'U+2212',
        description: 'MINUS SIGN',
        write: '−',
        label: '−',
      },
    ],
    [
      {
        unicode: 'U+200A',
        description: 'HAIR SPACE',
        write: String.fromCodePoint('0x200A'),
        label: '&hairsp;',
      },
      {
        unicode: 'U+2022',
        description: 'BULLET',
        write: '•',
        label: '•',
      },
      {
        unicode: 'U+0060',
        description: 'GRAVE ACCENT',
        write: '`',
        label: '`',
        extraProps: { style: { gridRow: 'span 2' } },
      },
      {
        unicode: 'U+00E0',
        description: 'LATIN SMALL LETTER A WITH GRAVE',
        write: 'à',
        label: 'à',
      },
      {
        unicode: 'U+00E8',
        description: 'LATIN SMALL LETTER E WITH GRAVE',
        write: 'è',
        label: 'è',
      },
      {
        unicode: 'U+00EC',
        description: 'LATIN SMALL LETTER I WITH GRAVE',
        write: 'ì',
        label: 'ì',
      },
      {
        unicode: 'U+00F2',
        description: 'LATIN SMALL LETTER O WITH GRAVE',
        write: 'ò',
        label: 'ò',
      },
      {
        unicode: 'U+00F9',
        description: 'LATIN SMALL LETTER U WITH GRAVE',
        write: 'ù',
        label: 'ù',
      },
      {
        unicode: 'U+2013',
        description: 'EN DASH',
        write: '–',
        label: '–',
      },
    ],
    [
      {
        unicode: 'U+0009',
        description: 'TAB',
        write: String.fromCodePoint('0x0009'),
        label: 'TAB',
      },
      {
        unicode: 'U+25E6',
        description: 'WHITE BULLET',
        write: '◦',
        label: '◦',
      },
      {
        unicode: 'U+00C0',
        description: 'LATIN CAPITAL LETTER A WITH GRAVE',
        write: 'À',
        label: 'À',
      },
      {
        unicode: 'U+00C8',
        description: 'LATIN CAPITAL LETTER E WITH GRAVE',
        write: 'È',
        label: 'È',
      },
      {
        unicode: 'U+00CC',
        description: 'LATIN CAPITAL LETTER I WITH GRAVE',
        write: 'Ì',
        label: 'Ì',
      },
      {
        unicode: 'U+00D2',
        description: 'LATIN CAPITAL LETTER O WITH GRAVE',
        write: 'Ò',
        label: 'Ò',
      },
      {
        unicode: 'U+00D9',
        description: 'LATIN CAPITAL LETTER U WITH GRAVE',
        write: 'Ù',
        label: 'Ù',
      },
      {
        unicode: 'U+2014',
        description: 'EM DASH',
        write: '—',
        label: '—',
      },
    ],
  ],
};

export const characterIcons = {
  spanish: 'ñ',
  special: '€',
};
