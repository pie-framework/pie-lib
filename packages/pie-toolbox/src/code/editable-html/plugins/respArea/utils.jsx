import React from 'react';
import ReactDOM from 'react-dom';
import Snackbar from '@material-ui/core/Snackbar';

export const isNumber = (val) => !isNaN(parseFloat(val)) && isFinite(val);

export const insertSnackBar = (message) => {
  const prevSnacks = document.querySelectorAll('.response-area-alert');

  prevSnacks.forEach((s) => s.remove());

  const newEl = document.createElement('div');

  newEl.className = 'response-area-alert';

  const el = (
    <Snackbar
      anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
      open={true}
      ContentProps={{
        'aria-describedby': 'message-id',
      }}
      message={<span id="message-id">{message}</span>}
    />
  );

  ReactDOM.render(el, newEl);

  document.body.appendChild(newEl);

  setTimeout(() => {
    newEl.remove();
  }, 2000);
};

export const defaultECR = (index) => ({
  type: 'explicit_constructed_response',
  children: [{ text: '' }],
  data: {
    index,
  },
});

export const defaultDIB = (opts, index) => ({
  type: 'drag_in_the_blank',
  children: [{ text: '' }],
  data: {
    index,
    duplicates: opts.options.duplicates,
    value: null,
  },
});

export const defaultIDD = (index) => ({
  type: 'inline_dropdown',
  children: [{ text: '' }],
  data: {
    index,
  },
});

export const getDefaultElement = (opts, index) => {
  switch (opts.type) {
    case 'explicit-constructed-response':
      return defaultECR(index);

    case 'drag-in-the-blank':
      return defaultDIB(opts, index);

    default:
      // inline-dropdown
      return defaultIDD(index);
  }
};
