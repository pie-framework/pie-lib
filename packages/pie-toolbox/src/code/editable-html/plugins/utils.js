export const findSingleNode = (value) => {
  if (!value || !value.isCollapsed || !value.startKey) {
    return;
  }

  const inline = value.document.getClosestInline(value.startKey);

  if (inline) {
    return inline;
  }

  const block = value.document.getClosestBlock(value.startKey);

  if (block) {
    return block;
  }
};

export const findParentNode = (value, node) => {
  if (!value || !node) {
    return;
  }

  return value.document.getParent(node.key);
};

export const hasMark = (value, type) => value && value.marks.some((mark) => mark.type == type);

export const hasBlock = (value, type) => value && value.blocks.some((node) => node.type == type);

export const hasNode = ({ document }, type) => document && document.nodes.some((node) => node.type == type);

let activeKeypad = null;

export const setActiveKeypad = (keypad) => {
  if (activeKeypad && activeKeypad !== keypad) {
    activeKeypad.close();
  }
  activeKeypad = keypad;
};

export const closeActiveKeypad = () => {
  if (activeKeypad) {
    activeKeypad.close();
    activeKeypad = null;
  }
};
