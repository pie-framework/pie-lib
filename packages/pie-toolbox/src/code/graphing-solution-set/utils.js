import head from 'lodash/head';
import tail from 'lodash/tail';
import { utils } from '../plot';
import invariant from 'invariant';
import isEqual from 'lodash/isEqual';
import isEmpty from 'lodash/isEmpty';
import cloneDeep from 'lodash/cloneDeep';

export const bounds = utils.bounds;
export const point = utils.point;

//TODO: This can be removed?
export const getAngleDeg = () => 0;
//TODO: This can be removed?
export const arrowDimensions = () => 0;

export const getTickValues = (prop) => {
  const tickValues = [];
  let tickVal = 0;

  while (tickVal >= prop.min && tickValues.indexOf(tickVal) < 0) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal - prop.step) * 10000) / 10000;
  }

  tickVal = Math.round(prop.step * 10000) / 10000;

  while (tickVal <= prop.max && tickValues.indexOf(tickVal) < 0) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal + prop.step) * 10000) / 10000;
  }

  // return only ticks that are inside the min-max interval
  if (tickValues) {
    return tickValues.filter((tV) => tV >= prop.min && tV <= prop.max);
  }

  return [];
};

export const countWords = (label) => {
  if (label == null || isEmpty(label)) {
    return 1;
  }

  const words = label.split(' ');
  return words.length;
};

// findLongestWord is also used in plot
export const findLongestWord = (label) => {
  let longestWord = (label || '')
    .replace(/<[^>]+>/g, '')
    .split(' ')
    .sort((a, b) => b.length - a.length);

  return longestWord[0].length;
};

// amountToIncreaseWidth is also used in plot
export const amountToIncreaseWidth = (longestWord) => {
  if (!longestWord) {
    return 0;
  }

  return longestWord * 10;
};

export const polygonToArea = (points) => {
  const h = head(points);
  const area = {
    left: h.x,
    top: h.y,
    bottom: h.y,
    right: h.x,
  };
  return tail(points).reduce((a, p) => {
    a.left = Math.min(a.left, p.x);
    a.top = Math.max(a.top, p.y);
    a.bottom = Math.min(a.bottom, p.y);
    a.right = Math.max(a.right, p.x);
    return a;
  }, area);
};

export const lineToArea = (from, to) => pointsToArea(from, to);

export const pointsToArea = (a, b) => {
  invariant(!!a && !!b, 'a or b is undefined');
  const left = Math.min(a.x, b.x);
  const top = Math.max(a.y, b.y);
  const bottom = Math.min(a.y, b.y);
  const right = Math.max(a.x, b.x);
  return { left, top, bottom, right };
};

export const getRightestPoints = (points) => {
  const sortedPoints = cloneDeep(points);
  sortedPoints.sort((a, b) => b.x - a.x);

  return { a: sortedPoints[0], b: sortedPoints[1] };
};

export const getMiddleOfTwoPoints = (a, b) => ({
  x: (a.x + b.x) / 2,
  y: (a.y + b.y) / 2,
});

export const roundNumber = (number) => parseFloat(number.toFixed(4));

export const sameAxes = (p1, p2) =>
  p1 && p2 && (roundNumber(p1.x) === roundNumber(p2.x) || roundNumber(p1.y) === roundNumber(p2.y));

export const equalPoints = (p1, p2) =>
  p1 &&
  p2 &&
  isEqual(
    {
      x: roundNumber(p1.x),
      y: roundNumber(p1.y),
    },
    {
      x: roundNumber(p2.x),
      y: roundNumber(p2.y),
    },
  );

const getDistanceBetweenTicks = (axis, size) => {
  const { min, max, step } = axis;
  const nbOfTicks = (max - min) / step;

  return size / nbOfTicks;
};

export const thinnerShapesNeeded = (graphProps) => {
  const {
    domain,
    range,
    size: { width, height },
  } = graphProps;

  // 14 is the default width of a point
  return getDistanceBetweenTicks(domain, width) < 14 || getDistanceBetweenTicks(range, height) < 14;
};

export const getAdjustedGraphLimits = (graphProps) => {
  const {
    domain,
    range,
    size: { width, height },
  } = graphProps;
  const domainTicksDistance = getDistanceBetweenTicks(domain, width);
  const rangeTicksDistance = getDistanceBetweenTicks(range, height);

  // 15 is the distance required for the arrow to extend the graph
  const domainPadding = domain.step / (domainTicksDistance / 15);
  const rangePadding = range.step / (rangeTicksDistance / 15);

  return {
    domain: {
      min: domain.min - domainPadding,
      max: domain.max + domainPadding,
    },
    range: {
      min: range.min - rangePadding,
      max: range.max + rangePadding,
    },
  };
};

const sortPoints = (array) => (array || []).sort((a, b) => a.x - b.x || a.y - b.y);

// check colliniarity of 3 points (source: https://www.geeksforgeeks.org/program-check-three-points-collinear/)
const checkCollinearity = (a, b, c) => (a.x - b.x) * (c.y - b.y) === (c.x - b.x) * (a.y - b.y);

// 2 lines are overlapping if all 4 points are collinear
const isSameLine = (markA, markB) =>
  checkCollinearity(markA.from, markB.from, markB.to) && checkCollinearity(markA.to, markB.from, markB.to);

export const isDuplicatedMark = (mark, marks, oldMark) => {
  const { type, building } = mark;

  if (building) {
    return false;
  }

  const filteredMarks = (marks || []).filter((m) => m.type === type && !m.building);
  const index = filteredMarks.findIndex((m) => isEqual(m, oldMark));

  if (index !== -1) {
    filteredMarks.splice(index, 1);
  }

  const duplicated = filteredMarks.find((m) => {
    if (type === 'line') {
      const { from, to } = mark;
      return (
        (equalPoints(from, m.from) && equalPoints(to, m.to)) ||
        (equalPoints(from, m.to) && equalPoints(to, m.from)) ||
        (type === 'line' && isSameLine(m, mark))
      );
    } else if (type === 'polygon') {
      return isEqual(sortPoints(cloneDeep(mark.points)), sortPoints(cloneDeep(m.points)));
    }
  });

  return !!duplicated;
};

export const areArraysOfObjectsEqual = (array1, array2) => {
  // Check if both arrays have the same length
  if (array1.length !== array2.length) {
    return false;
  }
  // Iterate through each object in the arrays
  for (let i = 0; i < array1.length; i++) {
    // Get the current objects in both arrays
    const obj1 = array1[i];
    const obj2 = array2[i];
    // Check if the objects have the same number of properties
    if (Object.keys(obj1).length !== Object.keys(obj2).length) {
      return false;
    }
    // Iterate through each property in the objects
    for (const key in obj1) {
      // Check if the properties and their values are equal
      if (obj1[key] !== obj2[key]) {
        return false;
      }
    }
  }
  // If all objects are equal, the arrays are the same
  return true;
};
