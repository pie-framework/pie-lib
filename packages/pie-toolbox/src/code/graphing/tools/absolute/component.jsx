import debug from 'debug';
import { buildDataPoints, absoluteFromTwoPoints } from '../../../graphing-utils';
import { withRootEdge, rootEdgeComponent } from '../shared/line/with-root-edge';

const log = debug('pie-lib:graphing:absolute');

const Absolute = withRootEdge((props) => {
  const { root, edge, graphProps } = props;
  const { domain, range } = graphProps;

  const dataPoints =
    edge && edge.x === root.x
      ? []
      : buildDataPoints(domain, range, root, edge, absoluteFromTwoPoints(root, edge), true);

  log('dataPoints:', dataPoints);

  return { root: props.root, edge: props.edge, dataPoints, enableCurve: false };
});

const Component = rootEdgeComponent(Absolute);

export default Component;
