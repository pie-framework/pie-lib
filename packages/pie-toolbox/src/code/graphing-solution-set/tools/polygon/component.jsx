import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { ToolPropTypeFields } from '../shared/types';
import chunk from 'lodash/chunk';
import initial from 'lodash/initial';
import debug from 'debug';
import DraggablePolygon from './polygon';
import { types } from '../../../plot';
import invariant from 'invariant';
import ReactDOM from 'react-dom';
import MarkLabel from '../../mark-label';
import isEmpty from 'lodash/isEmpty';
import { getMiddleOfTwoPoints, getRightestPoints, equalPoints } from '../../utils';

const log = debug('pie-lib:graphing-solution-set:polygon');

export const buildLines = (points, closed) => {
  const expanded = points.reduce((acc, p, index) => {
    acc.push({ ...p, index });

    const isLast = index === points.length - 1;
    const next = isLast ? 0 : index + 1;

    acc.push({ ...points[next], index: next });

    return acc;
  }, []);

  const all = chunk(expanded, 2).map(([from, to]) => {
    return { from, to };
  });

  return closed ? all : initial(all);
};

export const swap = (arr, ...rest) => {
  const pairs = chunk(rest, 2);

  return pairs.reduce(
    (acc, pr) => {
      if (pr.length === 2) {
        let [e, replacement] = pr;

        invariant(Number.isFinite(e.index), 'Index must be defined');

        const index = e.index;
        // const i = acc.findIndex(pt => pt.x === e.x && pt.y === e.y);

        if (index >= 0) {
          acc.splice(index, 1, replacement);

          return acc;
        } else {
          return acc;
        }
      } else {
        return acc;
      }
    },
    [...arr],
  );
};

export class RawBaseComponent extends React.Component {
  static propTypes = {
    classes: PropTypes.object,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    correctness: PropTypes.string,
    points: PropTypes.arrayOf(types.PointType),
    closed: PropTypes.bool,
    isSolution: PropTypes.bool,
    coordinatesOnHover: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    onClosePolygon: PropTypes.func.isRequired,
    onDragStart: PropTypes.func,
    onDragStop: PropTypes.func,
    onClick: PropTypes.func,
    graphProps: types.GraphPropsType.isRequired,
    isToolActive: PropTypes.bool,
    middle: PropTypes.object,
    labelNode: PropTypes.object,
    labelModeEnabled: PropTypes.bool,
    onChangeLabelProps: PropTypes.func,
    onChangeProps: PropTypes.func,
  };

  static defaultProps = {
    points: [],
  };

  dragPoint = (index, from, to) => {
    log('[dragPoint] from, to:', from, to);
    const { onChange, points } = this.props;
    const update = [...points];
    const overlapPoint = !!(points || []).find((p) => equalPoints(p, to));

    if (equalPoints(from, to) || overlapPoint) {
      return;
    }

    if (points[index].label) {
      to.label = points[index].label;
    }

    update.splice(index, 1, to);
    onChange(update);
  };

  dragLine = (existing, next) => {
    log('[dragLine]: ', existing, next);
    const { onChange } = this.props;

    if (existing.from.label) {
      next.from.label = existing.from.label;
    }

    if (existing.to.label) {
      next.to.label = existing.to.label;
    }

    const points = swap(this.props.points, existing.from, next.from, existing.to, next.to);
    onChange(points);
  };

  dragPoly = (existing, next) => {
    log('[dragPoly] ', existing, next);
    const { onChange } = this.props;

    next.forEach((point, index) => {
      if (existing[index].label) {
        next[index].label = existing[index].label;
      }
    });

    onChange(next);
  };

  close = () => {
    const { points, onClosePolygon } = this.props;
    log('[close] ...');

    if (points.length >= 3) {
      onClosePolygon();
    } else {
      log('[close] - nope');
    }
  };

  labelChange = (point, index) => {
    const { points, onChangeProps } = this.props;
    const updatedPoint = { ...point };

    if (!point.label || isEmpty(point.label)) {
      delete updatedPoint.label;
    }

    const update = [...points];

    update.splice(index, 1, updatedPoint);
    onChangeProps(update);
  };

  clickPoint = (point, index, data) => {
    const {
      closed,
      disabled,
      onClick,
      isToolActive,
      labelModeEnabled,
      onChangeProps,
      onChangeLabelProps,
      points,
    } = this.props;

    if (labelModeEnabled) {
      if (disabled) {
        return;
      }

      if (points && index === points.length) {
        const { a, b } = getRightestPoints(points);
        const middle = { label: '', ...point, ...getMiddleOfTwoPoints(a, b) };

        onChangeLabelProps(middle);
      } else {
        const update = [...points];

        update.splice(index, 1, { label: '', ...point });
        onChangeProps(update);
      }

      if (this.input[index]) {
        this.input[index].focus();
      }

      return;
    }

    if (isToolActive && !closed && index === 0) {
      this.close();
      return;
    }

    onClick(point || data);
  };

  // IMPORTANT, do not remove
  input = {};

  render() {
    const {
      closed,
      graphProps,
      onChangeLabelProps,
      points,
      middle,
      labelNode,
      labelModeEnabled,
      isSolution = false,
    } = this.props;
    const polygonLabelIndex = (points && points.length) || 0;
    if (labelNode && middle && middle.hasOwnProperty('label')) {
      ReactDOM.createPortal(
        <MarkLabel
          inputRef={(r) => (this.input[polygonLabelIndex] = r)}
          disabled={!labelModeEnabled}
          mark={middle}
          graphProps={graphProps}
          onChange={(label) => onChangeLabelProps({ ...middle, label })}
        />,
        labelNode,
      );
    }
    return (
      <g>
        <React.Fragment>
          <DraggablePolygon
            points={points}
            closed={closed}
            isSolution={isSolution}
            graphProps={graphProps}
            onClick={this.clickPoint.bind(this, middle, polygonLabelIndex)}
          />
        </React.Fragment>
      </g>
    );
  }
}

export const BaseComponent = withStyles(() => ({}))(RawBaseComponent);

export default class Component extends React.Component {
  static propTypes = {
    ...ToolPropTypeFields,
    graphProps: types.GraphPropsType.isRequired,
  };

  static defaultProps = {};

  constructor(props) {
    super(props);
    this.state = {};
  }

  change = (points) => {
    const {
      mark: { middle },
    } = this.props;
    const mark = { ...this.state.mark, points };

    if (middle) {
      const { a, b } = getRightestPoints(points);

      mark.middle = { ...middle, ...getMiddleOfTwoPoints(a, b) };
    }

    this.setState({ mark });
  };

  changeProps = (points) => {
    const mark = { ...this.props.mark, points };

    this.props.onChange(this.props.mark, mark);
  };

  changeLabelProps = (point) => {
    const { mark, onChange } = this.props;
    const middle = { ...mark.middle, ...point };

    onChange(mark, { ...mark, middle });
  };

  closePolygon = () => {
    log('[closePolygon] ...');
    const { onComplete, mark } = this.props;
    const update = { ...mark, closed: true };

    onComplete(mark, update);
  };

  dragStart = () => this.setState({ mark: this.props.mark });

  dragStop = () => {
    const { onChange } = this.props;
    const m = { ...this.state.mark };

    this.setState({ mark: undefined }, () => {
      onChange(this.props.mark, m);
    });
  };

  render() {
    const { coordinatesOnHover, mark, graphProps, onClick, isToolActive, labelNode, labelModeEnabled } = this.props;
    const { mark: stateMark } = this.state;

    return (
      <BaseComponent
        {...(stateMark || mark)}
        coordinatesOnHover={coordinatesOnHover}
        onChange={this.change}
        onChangeLabelProps={this.changeLabelProps}
        onChangeProps={this.changeProps}
        onClosePolygon={this.closePolygon}
        onDragStart={this.dragStart}
        onDragStop={this.dragStop}
        onClick={onClick}
        graphProps={graphProps}
        isToolActive={isToolActive}
        labelNode={labelNode}
        labelModeEnabled={labelModeEnabled}
      />
    );
  }
}
