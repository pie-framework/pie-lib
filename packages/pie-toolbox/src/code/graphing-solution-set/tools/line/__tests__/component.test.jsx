import { shallow } from 'enzyme';
import React from 'react';
import { ArrowedLine } from '../component';
import { graphProps as getGraphProps } from '../../../__tests__/utils';
jest.mock('../../../../plot/index', () => {
  const a = jest.requireActual('../../../../plot/index');
  return {
    types: a.types,
    gridDraggable: a.gridDraggable,
    utils: a.utils,
    trig: {
      edges: jest.fn(() => jest.fn().mockReturnValue([0, 0])),
    },
  };
});

describe('ArrowedLine', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = (extras) => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      markerId: '1',
      graphProps: getGraphProps(),
    };
    const props = { ...defaults, ...extras };
    return shallow(<ArrowedLine {...props} />);
  };
  describe('snapshot', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });
});
