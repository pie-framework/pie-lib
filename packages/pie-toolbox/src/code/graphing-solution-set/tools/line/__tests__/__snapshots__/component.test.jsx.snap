// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowedLine snapshot renders 1`] = `
<g>
  <defs>
    <ArrowMarker
      className=""
      id="1-enabled"
      size={5}
    />
  </defs>
  <line
    className="className"
    markerEnd="url(#1-enabled)"
    markerId="1"
    markerStart="url(#1-enabled)"
    onChange={[MockFunction]}
  />
</g>
`;
